Compile/src/Producers/PUJetIDWeightProducer.cc

```diff
@@ -0,0 +1,85 @@
+#include "Excalibur/Compile/interface/Producers/PUJetIDWeightProducer.h"
+#include "KappaTools/RootTools/interface/Matching.h"
+
+std::string PUJetIDWeightProducer::GetProducerId() const { return "PUJetIDWeightProducer"; }
+
+void PUJetIDWeightProducer::Init(ZJetSettings const& settings) {
+
+    PUJetID = settings.GetPUJetID();
+    PUJetIDSelectionMaxPt = 50.0;
+    inputIsData = settings.GetInputIsData();
+
+    if (inputIsData)
+    {
+        LOG(FATAL) << "PUJetID SFs can only be applied to MC!";
+    }
+
+    if (PUJetID != "none")
+    {
+        LOG(INFO) << "Apply PUJetID cut on jets pt<" << PUJetIDSelectionMaxPt << " GeV";
+        PUEffFilename = settings.GetPUJetIDEffFilename();
+        std::string histName = settings.GetPUJetIDEffHistogramName();
+        TFile PUEffFile(PUEffFilename.c_str(), "READONLY");
+        if (PUEffFile.IsOpen() == false)
+        {
+            LOG(FATAL) << "File for PUJetID scalefactors (" << PUEffFilename <<
+                " not found. Please check PUEffFilename.";
+        }
+
+        LOG(INFO) << "load histogram " << histName << " for PUJetID scalefactors";
+        sfhisto = (TH2F*)PUEffFile.Get(histName.c_str());
+        if (!sfhisto)
+        {
+            LOG(FATAL)<<"no histogram found with name" << histName;
+        }
+        sfhisto->SetDirectory(0);
+
+        std::string errName = histName+"_Systuncty";
+        LOG(INFO) << "load histogram " << errName << " for PUJetID scalefactors uncertainties.";
+        errhisto = (TH2F*)PUEffFile.Get(errName.c_str());
+        if (!errhisto)
+        {
+            LOG(FATAL)<<"no uncertainty histogram found with name" << errName;
+        }
+        errhisto->SetDirectory(0);
+
+        PUEffFile.Close();
+    }
+}
+
+void PUJetIDWeightProducer::Produce(ZJetEvent const& event,
+                                    ZJetProduct& product,
+                                    ZJetSettings const& settings) const {
+
+    LOG(DEBUG) << "\n[" << this->GetProducerId() << "]";
+    product.m_optionalWeights["puJetIDWeight"] = 1.0;
+    product.m_optionalWeights["puJetIDWeightUp"] = 1.0;
+    product.m_optionalWeights["puJetIDWeightDown"] = 1.0;
+
+    const std::vector<std::shared_ptr<KJet>> theJets = product.m_correctedZJets["L1L2L3"];
+    const std::vector<int> matchedGenJetIndices = product.m_matchedGenJets4["L1L2L3"];
+    const KLV* matchedGenJet = nullptr;
+    if (matchedGenJetIndices.size() < 1 || theJets.size() < 1) {
+        LOG(DEBUG) << "No Jets and/or GenJets available.";
+        return;
+    }
+    if (matchedGenJetIndices[0] >= 0) {
+        matchedGenJet = product.m_simpleGenJets[matchedGenJetIndices[0]];
+    }
```

Comment:
When I understand it correctly, this is in order to check only for the leading jet. When you want to change this strategy, it might be beneficial to extend the code.

Reply:
Right now we only cut event wise on requirements of the leading jet. This of course would be needed, if we have requirements on more than the leading jet. This requires some thoughts on how to handle this.

Reply:
So after discussion, it just has to be ensured, that this producer always runs after all jet vetoes. That would be fine, I guess.

However, just for the sake of completion, another option, which defaults to `1`, would be nice to have, in case one wants to select more than one jet.

Reply:
Looking at the twiki, it states the SF should be applied on all Jets in an event, so I updated the behaviour to do exactly this. Since genuine 2 or more jet events are rare comared to 1 jet events, this shouldn't influence the final results by much.